### PR TITLE
Upgrade to arrow 20.0.0

### LINKED ISSUE #3006: The upcoming release of Arrow (20?) breaks datafusion
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Some changes to Decimal 128 & 256 require updating datafusion code to support.

**Describe the solution you'd like**

Decimal 128 to be supported.

**Describe alternatives you've considered**

I think we're going to have to do this.

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by avantgardnerio**
# Which issue does this PR close?

Closes #3006.

 # Rationale for this change

We'll have to do it sooner or later, but mostly I want the new authentication capabilities in the `FlightSqlServer` code.

# What changes are included in this PR?

An attempt at upgrading to the latest arrow.

# Are there any user-facing changes?

It doesn't (presently) compile, so I'm making a draft PR to seek input on CI failures.

### CODE REVIEW

**Comment by avantgardnerio** [comment]
I fixed a lot of the obvious errors, but I wasn't sure what to do about
```
error[E0308]: mismatched types
   --> datafusion/physical-expr/src/expressions/binary.rs:135:40
    |
135 |         .map(|left| left.map(|left| op(left, right)))
    |                                     -- ^^^^ expected `i128`, found struct `arrow::util::decimal::Decimal128`
    |                                     |
    |                                     arguments to this function are incorrect
    |
note: callable defined here
   --> datafusion/physical-expr/src/expressions/binary.rs:131:8
    |
131 |     F: Fn(i128, i128) -> bool,
    |        ^^^^^^^^^^^^^^^^^^^^^^
```
and
```
error[E0308]: mismatched types
    --> datafusion/physical-expr/src/expressions/in_list.rs:225:68
     |
225  |                     .map(|vop| match vop.map(|v| !$VALUES.contains(&v)) {
     |                                                           -------- ^^ expected `i128`, found struct `arrow::util::decimal::Decimal128`
     |                                                           |
     |                                                           arguments to this function are incorrect
...
318  |     collection_contains_check!(array, values, negated, contains_null)
     |     ----------------------------------------------------------------- in this macro invocation
     |
     = note: expected reference `&i128`
                found reference `&arrow::util::decimal::Decimal128`
note: associated function defined here
    --> /home/bgardner/snap/rustup/common/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/mod.rs:2172:12
     |
2172 |     pub fn contains(&self, x: &T) -> bool
     |            ^^^^^^^^
     = note: this error originates in the macro `collection_contains_check` (in Nightly builds, run with -Z macro-backtrace for more info)
```
CC @andygrove and @alamb  - if you guys have any advice, it would be appreciated.

**Comment by avantgardnerio** [comment]
If I understand correctly, the other XXXArray types have iterators that return the primitive representations (i.e. `u128`), but Decimal now returns a struct with `From` and `to()` traits?

https://github.com/apache/arrow-rs/blob/master/arrow/src/util/decimal.rs#L132

**Comment by avantgardnerio** [comment]
If I revert the change here: https://github.com/apache/arrow-rs/pull/2140/files to:

```
def_decimal_array!(
    Decimal128Array,
    "Decimal128Array",
    Decimal128,
    Decimal128Iter
);
```
vs
```
def_decimal_array!(Decimal128Array, "Decimal128Array");
```
It compiles and happily runs. I'm not sure if that's the correct fix though.

**Comment by avantgardnerio** [comment]
CC @HaoYang670

**Comment by avantgardnerio** [comment]
I think we need to just change all these: https://github.com/apache/arrow-datafusion/blob/55a12869f32644ed87d69d9b7617372db5103fb8/datafusion/physical-expr/src/expressions/binary.rs#L260

From `i128` to `Decimal128`

**Comment by avantgardnerio** [comment]
> From i128 to Decimal128

But that ripples up to ScalarValues and throughout the codebase quickly.

**Comment by viirya** [comment]
> expected `i128`, found struct `arrow::util::decimal::Decimal128`

Have you tried calling `as_i128()`?

**Comment by stuartcarnie** [comment]
👋🏻 hi @avantgardnerio I was able to successfully upgrade `arrow-rs` via this draft PR for which all DataFusion tests pass: https://github.com/apache/arrow-datafusion/pull/3010. Perhaps there are commits in this PR that are useful for your work here?

**Comment by alamb** [comment]
Thank you @avantgardnerio and @stuartcarnie  -- this is so great. I have normally been doing this work during / as part of the release process (e.g. https://github.com/apache/arrow-rs/issues/2172) 

THANK YOU so much for getting a start on it earlier.

**Comment by avantgardnerio** [comment]
> Perhaps there are commits in this PR that are useful for your work here?

100%. Thank you so much!

**Comment by avantgardnerio** [comment]
> Have you tried calling `as_i128()`?

I did, but however I tried to do it rippled up through the codebase and affected lots of places. @andygrove was able to figure it out. Thank you both for your help!

**Comment by avantgardnerio** [comment]
@andygrove and @alamb looks like the checks are failing because lack of `protoc`? Any idea why this would have just started?

**Comment by carols10cents** [comment]
Yes, prost is no longer bundling protoc. It looks like this repo will need changes similar to these? https://github.com/apache/arrow-rs/pull/2280

**Comment by alamb** [comment]
@avantgardnerio I can help mess around with (aka help with) this PR as part of preparing the arrow release candidate, which I will likely do this Friday

**Comment by avantgardnerio** [comment]
> likely do this Friday

I'd love to get this one in first, but I'm sure there's a lot of people with a lot of interests and they all need to be considered: https://github.com/apache/arrow-rs/pull/2309

I'm happy to help in any way I can with the release.

**Comment by avantgardnerio** [comment]
@tustvold I think I'm stuck on these failures:

```
---- physical_plan::file_format::parquet::tests::parquet_exec_with_error stdout ----
thread 'physical_plan::file_format::parquet::tests::parquet_exec_with_error' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidPath { path: "." }', datafusion/core/src/physical_plan/file_format/parquet.rs:1243:14

---- test_util::tests::test_data_dir stdout ----
thread 'test_util::tests::test_data_dir' panicked at 'assertion failed: `(left == right)`
  left: `"/Users/runner/work/arrow-datafusion/arrow-datafusion/datafusion"`,
 right: `"/Users/runner/work/arrow-datafusion/arrow-datafusion/datafusion/core/.."`', datafusion/core/src/test_util.rs:345:9
```

It appears they are because this branch is referencing `../objectstore` instead of `object_store = "0.3.0"` on master. I'm a little reticent to switch is back as this seems more appropriate and there's build errors either way. When you get a moment, I'd appreciate input, thanks!

**Comment by alamb** [comment]
> It appears they are because this branch is referencing ../objectstore instead of object_store = "0.3.0" on master. I'm a little reticent to switch is back as this seems more appropriate and there's build errors either way. When you get a moment, I'd appreciate input, thanks!

I'll try and give it a look on monday. Thanks @avantgardnerio

File: datafusion/common/Cargo.toml [hunk 1]
```diff
@@ -39,11 +39,12 @@ pyarrow = ["pyo3"]
 
 [dependencies]
 apache-avro = { version = "0.14", features = ["snappy"], optional = true }
-arrow = { version = "19.0.0", features = ["prettyprint"] }
+arrow = { git = "https://github.com/apache/arrow-rs.git", rev = "30c94dbf1c422f81f8520b9956e96ab7b53c3f47", features = ["prettyprint"], optional = false }
+avro-rs = { version = "0.13", features = ["snappy"], optional = true }
 cranelift-module = { version = "0.86.1", optional = true }
-object_store = { version = "0.3", optional = true }
+object_store = { git = "https://github.com/apache/arrow-rs.git", rev = "30c94dbf1c422f81f8520b9956e96ab7b53c3f47", features = [], optional = true }
```
Comment (tustvold) [inline datafusion/common/Cargo.toml:45]: I don't believe we intend to release object_store as part of arrow 20, and so I would back out this change

File: datafusion/core/Cargo.toml [hunk 2]
```diff
@@ -75,10 +75,10 @@ lazy_static = { version = "^1.4.0" }
 log = "^0.4"
 num-traits = { version = "0.2", optional = true }
 num_cpus = "1.13.0"
-object_store = "0.3.0"
+object_store = { git = "https://github.com/apache/arrow-rs.git", rev = "30c94dbf1c422f81f8520b9956e96ab7b53c3f47", features = [], optional = false }
```
Comment (tustvold) [inline datafusion/core/Cargo.toml:78]: Ditto

**Comment by avantgardnerio** [comment]
Thanks @alamb and @tustvold . I'll be out most of today, I'll try to devote time as I have it this weekend. I'm sure we can get it sorted by Monday.

File: datafusion/common/Cargo.toml (hunk 1 above)
Comment (avantgardnerio) [reply datafusion/common/Cargo.toml:45]: Here's an alternative version with `object_store` not upgraded, but I struggled with this too: https://github.com/apache/arrow-datafusion/pull/3007

**Comment by avantgardnerio** [comment]
I got all the tests passing except windows. I installed a VM and tried to fix that too, but I'm getting lots of interesting errors with paths like `file:///?\C:\some\path/myfile.parquet` and I have to admit I'm a bit lost on those. I can probably debug them, but I might need to set up a real Windows environment instead of a slow VM.

**Comment by tustvold** [comment]
> I got all the tests passing except windows. I installed a VM and tried to fix that too, but I'm getting lots of interesting errors with paths like file:///?\C:\some\path/myfile.parquet and I have to admit I'm a bit lost on those. I can probably debug them, but I might need to set up a real Windows environment instead of a slow VM.

The path handling changed upstream, in particular https://github.com/apache/arrow-rs/pull/2269, I could definitely see this having downstream implications. There is an outstanding PR https://github.com/apache/arrow-rs/pull/2371 that may help, but I would strongly advise against upgrading object_store as part of this PR

**Comment by avantgardnerio** [comment]
@alamb & @andygrove the only remaining failure looks like an intermittent network issue. I think we're GTG here :tada:

**Comment by avantgardnerio** [comment]
Closing in favor of https://github.com/apache/arrow-datafusion/pull/3083

**Pull request closed without merging**